### DISCUSSION TITLE
Job is incomplete class Exception

### DISCUSSION THREAD

**Discussion by RaidOuahioune** [laravel/framework#49540] [Q&A]
### Laravel Version

10.10

### PHP Version

8.2

### Description

I'm using redis as a queue driver and manged by horizon. I have 10 queue workers .I'm randomly getting this error    "message": "Job is incomplete class: but when i retry the job next time it works fine.
Here is a detailed log 
{
    "class": "Exception",
    "message": "Job is incomplete class: {\"__PHP_Incomplete_Class_Name\":\"App\\\\Jobs\\\\Trends\\\\GlobalRecentsUpdater\",\"connection\":\"redis\",\"queue\":\"trends\"}",
    "code": 0,
    "file": "/var/www/vendor/laravel/framework/src/Illuminate/Queue/CallQueuedHandler.php:117",
    "trace": [
        "/var/www/vendor/laravel/framework/src/Illuminate/Queue/CallQueuedHandler.php:70",
        "/var/www/vendor/laravel/framework/src/Illuminate/Queue/Jobs/Job.php:102",
        "/var/www/vendor/laravel/framework/src/Illuminate/Queue/Worker.php:439",
        "/var/www/vendor/laravel/framework/src/Illuminate/Queue/Worker.php:389",
        "/var/www/vendor/laravel/framework/src/Illuminate/Queue/Worker.php:176",
        "/var/www/vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php:137",
        "/var/www/vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php:120",
        "/var/www/vendor/laravel/horizon/src/Console/WorkCommand.php:51",
        "/var/www/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php:36",
        "/var/www/vendor/laravel/framework/src/Illuminate/Container/Util.php:41",
        "/var/www/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php:93",
        "/var/www/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php:35",
        "/var/www/vendor/laravel/framework/src/Illuminate/Container/Container.php:662",
        "/var/www/vendor/laravel/framework/src/Illuminate/Console/Command.php:211",
        "/var/www/vendor/symfony/console/Command/Command.php:326",
        "/var/www/vendor/laravel/framework/src/Illuminate/Console/Command.php:180",
        "/var/www/vendor/symfony/console/Application.php:1096",
        "/var/www/vendor/symfony/console/Application.php:324",
        "/var/www/vendor/symfony/console/Application.php:175",
        "/var/www/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php:201",
        "/var/www/artisan:35

And here is my Job class  

```php
<?php

namespace App\Jobs\Scheduled\Users;

use App\Business\Config\GlobalUserSettings;
use App\Helpers\Cache\CacheKeys;
use App\Helpers\Cache\CacheLayer;
use App\Models\User;
use Illuminate\Bus\Queueable;
use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Foundation\Bus\Dispatchable;
use Illuminate\Queue\InteractsWithQueue;
use Illuminate\Queue\SerializesModels;

class GlobalUserReputationPointUpdater implements ShouldQueue
{
    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;


    public function handle(): void
    {
        User::query()
            ->whereIn("account_type", [0, 1])
            ->with('rawUserActivities')
            ->lazy(1000)
            ->each(function ($user) {
                $user->updateReputation();
            });
        $activeUsers = User::query()->orderBy('reputation_points', 'desc')->limit(GLobalUserSettings::ACTIVE_USERS_SIZE)->get();

        CacheLayer::CreateOrOverrideCachedData(CacheKeys::ACTIVE_USERS_KEY, $activeUsers, GLobalUserSettings::REMEMBER_ACTIVE_USERS);
    }
}
```

And Here is where i schedule the Job

```php
<?php

namespace App\Console;

use App\Helpers\Queues\QueueKeys;
use App\Jobs\Scheduled\Courses\CourseStartJob;
use App\Jobs\Scheduled\Invoices\ExpirationNotifier;
use App\Jobs\Scheduled\Invoices\MonthlyMoneyTransferJob;
use App\Jobs\Scheduled\Invoices\PremuimExpirationJob;
use App\Jobs\Scheduled\Trends\GlobalRecentsUpdater;
use App\Jobs\Scheduled\Trends\GlobalTrendsUpdater;
use App\Jobs\Scheduled\Trends\ReplieConfidenceUpdater;
use App\Jobs\Scheduled\Users\GeneratedRecommendations;
use App\Jobs\Scheduled\Users\GlobalUserReputationPointUpdater;
use App\Jobs\Scheduled\Users\UnSuspendUsers;
use Illuminate\Console\Scheduling\Schedule;
use Illuminate\Foundation\Console\Kernel as ConsoleKernel;

class Kernel extends ConsoleKernel
{
    /**
     * Define the application's command schedule.
     */
    protected function schedule(Schedule $schedule): void
    {
        $connection = 'redis';

        $schedule->job(new GlobalUserReputationPointUpdater, QueueKeys::REPUTATION_QUEUE, $connection)->hourly();
      
    }

    protected function commands(): void
    {
        $this->load(__DIR__ . '/Commands');

        require base_path('routes/console.php');
    }
}
```

### Steps To Reproduce

Actually I could not find any pattern for it

**Comment by leonardyrj** [comment]
I'm having the same problem, did I manage to solve it?

**Reply by RaidOuahioune** [reply]
Unfortunately :(   No I didn't.
I guess the problem is not with Laravel but with php itself.

**Comment by jordanade** [comment]
Also having this issue.
```
Exception: Job is incomplete class: {"__PHP_Incomplete_Class_Name":"App\\Jobs\\CheckSiteForBrokenLinks"} in /var/app/current/vendor/laravel/framework/src/Illuminate/Queue/CallQueuedHandler.php:117
Stack trace:
#0 /var/app/current/vendor/laravel/framework/src/Illuminate/Queue/CallQueuedHandler.php(70): Illuminate\Queue\CallQueuedHandler->dispatchThroughMiddleware(Object(Illuminate\Queue\Jobs\RedisJob), Object(__PHP_Incomplete_Class))
#1 /var/app/current/vendor/laravel/framework/src/Illuminate/Queue/Jobs/Job.php(102): Illuminate\Queue\CallQueuedHandler->call(Object(Illuminate\Queue\Jobs\RedisJob), Array)
#2 /var/app/current/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(439): Illuminate\Queue\Jobs\Job->fire()
#3 /var/app/current/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(389): Illuminate\Queue\Worker->process('redis', Object(Illuminate\Queue\Jobs\RedisJob), Object(Illuminate\Queue\WorkerOptions))
#4 /var/app/current/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(176): Illuminate\Queue\Worker->runJob(Object(Illuminate\Queue\Jobs\RedisJob), 'redis', Object(Illuminate\Queue\WorkerOptions))
#5 /var/app/current/vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php(138): Illuminate\Queue\Worker->daemon('redis', 'default', Object(Illuminate\Queue\WorkerOptions))
#6 /var/app/current/vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php(121): Illuminate\Queue\Console\WorkCommand->runWorker('redis', 'default')
#7 /var/app/current/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(36): Illuminate\Queue\Console\WorkCommand->handle()
#8 /var/app/current/vendor/laravel/framework/src/Illuminate/Container/Util.php(41): Illuminate\Container\BoundMethod::Illuminate\Container\{closure}()
#9 /var/app/current/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(93): Illuminate\Container\Util::unwrapIfClosure(Object(Closure))
#10 /var/app/current/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(37): Illuminate\Container\BoundMethod::callBoundMethod(Object(Illuminate\Foundation\Application), Array, Object(Closure))
#11 /var/app/current/vendor/laravel/framework/src/Illuminate/Container/Container.php(662): Illuminate\Container\BoundMethod::call(Object(Illuminate\Foundation\Application), Array, Array, NULL)
#12 /var/app/current/vendor/laravel/framework/src/Illuminate/Console/Command.php(211): Illuminate\Container\Container->call(Array)
#13 /var/app/current/vendor/symfony/console/Command/Command.php(326): Illuminate\Console\Command->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Illuminate\Console\OutputStyle))
#14 /var/app/current/vendor/laravel/framework/src/Illuminate/Console/Command.php(181): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Illuminate\Console\OutputStyle))
#15 /var/app/current/vendor/symfony/console/Application.php(1096): Illuminate\Console\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#16 /var/app/current/vendor/symfony/console/Application.php(324): Symfony\Component\Console\Application->doRunCommand(Object(Illuminate\Queue\Console\WorkCommand), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#17 /var/app/current/vendor/symfony/console/Application.php(175): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#18 /var/app/current/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(201): Symfony\Component\Console\Application->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#19 /var/app/current/artisan(37): Illuminate\Foundation\Console\Kernel->handle(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#20 {main}
```

**Comment by jordanade** [comment]
In my case, the issue turned out to be a separate QA server, connected to the same db/cache/queue, but with an older or nonexistent job class definition, that was attempting to run the queued jobs.

**Comment by martin-ro** [comment]
I'm having the same issue. Existing jobs still work just fine. I even copied a working job and renamed it with the same result:

```
Exception: Job is incomplete class: {"__PHP_Incomplete_Class_Name":"App\\Jobs\\
```

**Comment by e-rasekh** [comment]
I have this problem, too.

`Exception: Job is incomplete class: {"__PHP_Incomplete_Class_Name":"Modules\\Jobs\\SendJob"} `

**Comment by Elshaden** [comment]
Same here, same error.
Exception: Job is incomplete class: {"__PHP_Incomplete_Class ....

**Comment by codemountain** [comment]
Same here: Job is incomplete class: {"__PHP_Incomplete_Class_Name":"App\\Jobs

**Comment by vrajroham** [comment]
I encountered a similar issue and traced it back to the underlying classes/methods in use. The problem was related to the serialized class object stored in the jobs table/Redis.

When a job is dispatched, it is created and stored in the table as a serialized string using the [`serialize()`](https://www.php.net/manual/en/function.serialize.php) method.

When a queue worker picks up the jobs from the table/Redis, it unserializes the job object using the [`unserialize()`](https://www.php.net/manual/en/function.unserialize.php) method. In our case, different servers were processing different queues. One server dispatched the job, while another server with older code processed it. The serialized object class was not available on the second server.

Internally, the object is reconstructed from the serialized string using the [`unserialize()`](https://www.php.net/manual/en/function.unserialize.php) method. If the class implementation is not found, it replaces the class name with `__PHP_Incomplete_Class` and properties with `__PHP_Incomplete_Class_Name`, etc.

The issue seems to be that the classes are not found.

Here is the proof of concept: https://onecompiler.com/php/42j6zcceu

**Reply by ThatLadyDev** [reply]
You keep saying this, how were you able to resolve your issue?

**Reply by henzeb** [reply]
well, his problem was a outdated codebase in one of his servers. basically I think the solution here is to stop processing of jobs until you updated both servers. And don't add breaking changes to the code, that breaks deserialization.

**Reply by vrajroham** [reply]
The serialized object class was not available on the second server, itself answers the question that you need to check is that class is available in your code where queue workers are running.

**Although, here's a short summary about what was the issue and how I fixed it:**

The problem occurred because I removed one job class(E.g. `CalculateTrajectoryJob`) in a release, but that job was still queued in Redis for execution. When it was time for the job to run (after new release), it couldn’t be found, resulting in a `__PHP_Incomplete_Class` error. Also, even if you move the job to another namespace, the same error will occur.

To fix it, we added the removed class back to the code, and it worked. Going forward, we won’t remove jobs immediately. Instead, we’ll mark them as `@deprecated` and specify the date after which they can be safely removed from the codebase.

**Reply by RaidOuahioune** [reply]
I did not delete any kind of jobs in my code. My setup is as follows: a master server that runs horizon and connected to a redis server that all the other slaves are connected to . The slaves and master share exactly the same code

**Comment by liuzhaobo1999** [comment]
我也遇到了相同问题，已经解决了。
这个问题是队列执行时加载不了对应的类文件导致的。
一般执行完下面两部是没这个问题的：

1. 刷新代码缓存: php artisan optimize:clear
2. 重新加载队列: supervisorctl reload

我有多个服务器，项目的队列配置是同一个Redis 同一个库，但是有一个服务器B代码没有同步， 当队列存储到 Redis 后， 这些服务器都会从 Redis 队列里面拿到任务执行，服务器 B拿到后执行时没有找到最新代码，所以就报这个错。
大家可以排查一下

**Reply by henzeb** [reply]
Can we please use English here?

**Comment by Sampath-Wijesinghe** [comment]
Has anyone found the solution?

**Reply by henzeb** [reply]
Most of the time this happens when you changed the job signature. If you deployed it while old versions of the job are still waiting on the queue, you get this error.

If you need to change job signature, be sure the old jobs are completed or create a v2 instead.

**Comment by msiShariful** [comment]
Faced the same issue — it happens randomly. It seems to be a framework/PHP-level issue. I bypassed it using dispatchSync, which obviously doesn’t run the task in a separate process.


```
Message
Job is incomplete class: {"__PHP_Incomplete_Class_Name":"App\\Jobs\\JobName"}

Level
ERROR

Exception
{
    "class": "Exception",
    "message": "Job is incomplete class: {\"__PHP_Incomplete_Class_Name\":\"App\\\\Jobs\\\\JobName\"}",
    "code": 0,
    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Queue/CallQueuedHandler.php:116",
    "trace": [
        "/var/www/html/vendor/laravel/framework/src/Illuminate/Queue/CallQueuedHandler.php:69",
        "/var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Jobs/Job.php:102",
        "/var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Worker.php:442",
        "/var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Worker.php:392",
        "/var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Worker.php:335",
        "/var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php:149",
        "/var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php:132",
        "/var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php:36",
        "/var/www/html/vendor/laravel/framework/src/Illuminate/Container/Util.php:43",
        "/var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php:95",
        "/var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php:35",
        "/var/www/html/vendor/laravel/framework/src/Illuminate/Container/Container.php:696",
        "/var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php:213",
        "/var/www/html/vendor/symfony/console/Command/Command.php:279",
        "/var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php:182",
        "/var/www/html/vendor/symfony/console/Application.php:1094",
        "/var/www/html/vendor/symfony/console/Application.php:342",
        "/var/www/html/vendor/symfony/console/Application.php:193",
        "/var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php:198",
        "/var/www/html/vendor/laravel/framework/src/Illuminate/Foundati
```

**Comment by wlius-support3** [comment]
Had this issue yesterday - it was a question of psr-4 autoloading (in particular, just capitalization issues, which didn't exist on my dev machine (windows), but did in production (Linux)).
Here was the issue:
`Class App\Jobs\ActivateSIM located in ./app/Jobs/ActivateSim.php`

So I just needed to rename the `ActivateSim.php` to `ActivateSIM.php`. I found the issue when trying to manually call the job from a test file.

test.php:

```php
<?php

use App\Jobs\ActivateSIM;

// Bootstrap Laravel
require __DIR__.'/vendor/autoload.php';

$app = require_once __DIR__.'/bootstrap/app.php';

$kernel = $app->make(Illuminate\Contracts\Console\Kernel::class);
$kernel->bootstrap();

try {
    ActivateSIM::dispatch("1234", "4321");
} catch (Throwable $e) {
    echo 'Error: '.$e->getMessage().PHP_EOL;
    echo $e->getTraceAsString();
}
```

Running that test file showed me a clearer error message and that's how I found the PSR-4 autoloading issue.

Note, after fixing the file name, I optimized the autoloader like so:
`php /usr/local/bin/composer install --no-dev --optimize-autoloader`

That might also help you find any errors (as running it before fixing the file name also shows the PSR-4 issue).

**Comment by voidgraphics** [comment]
Ran into this today, it was because I have two sites on the same server that connect to the same database and I had queue workers on both.

I deployed a change on one of them, which restarted its queue worker, but the other queue worker was not restarted and was picking up the jobs which caused the error.

This made it pretty confusing because sometimes I would get the error and sometimes not, depending on which queue worker processed the job.

**Reply by Elshaden** [reply]
Fixed it by creating site-specific jobs, job_batches, and failed_jobs tables and setting them in the queue config and .env. This isolates each site’s jobs and prevents mix-ups.

**Comment by Sandalanka** [comment]
Why This Error Happens

That weird message:

Job is incomplete class: {"__PHP_Incomplete_Class_Name":"App\\Jobs\\Trends\\GlobalRecentsUpdater"}

means your queue worker pulled a job from Redis, but PHP couldn’t rebuild the original job object because the class definition wasn’t available or didn’t match the serialized data.

Think of it like this: Redis stored a “snapshot” of your job at the moment it was queued — including the class name and structure.
If, later, your code changes or workers are running outdated code, PHP goes “I don’t know what this class is” and marks it as __PHP_Incomplete_Class.

That’s what triggers this exception.

Why It’s Random

It feels random because Horizon runs multiple workers (in your case, 10).
Some workers might be using fresh code (after you deployed), while others are still using an older autoload map in memory.
So one worker unserializes the job fine, another can’t.

Retrying works because by that time, the right worker or updated process picks it up.

What Usually Causes It

Here’s the short list of suspects:

Code Deployment Without Worker Restart

Horizon workers keep running old code until you manually restart them.

Namespace or File Move

Maybe you moved or renamed the job class (GlobalRecentsUpdater) — Redis still has jobs pointing to the old path.

Cached Autoloader

Laravel’s cached class map or config might be stale after deployment.

Old Jobs Stuck in Redis

If you changed namespaces recently, Redis could still have old serialized jobs queued up.

How to Fix It Properly

Clear cached files before restarting Horizon:

php artisan optimize:clear


Terminate all Horizon workers so they reload fresh code:

php artisan horizon:terminate


(Don’t confuse this with queue:restart — Horizon ignores that one.)

If you renamed or moved job classes, flush the old jobs:

php artisan queue:flush


That clears Redis of any serialized jobs referencing old class names.

Switch to class-based scheduling instead of object-based:
Instead of:

$schedule->job(new GlobalUserReputationPointUpdater, QueueKeys::REPUTATION_QUEUE, 'redis')->hourly();


Use:

$schedule->job(GlobalUserReputationPointUpdater::class, QueueKeys::REPUTATION_QUEUE, 'redis')->hourly();


This avoids serializing the job instance itself into the scheduler’s cache.

Double-check class location:
Make sure the file exists:

app/Jobs/Trends/GlobalRecentsUpdater.php


and has:

namespace App\Jobs\Trends;
